File: Trinity.Runtime/platform_src/android/app/src/main/java/org/elastos/trinity/runtime/DIDSessionManagerPlugin.java
/*
 * Copyright (c) 2018 Elastos Foundation
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package org.elastos.trinity.runtime;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.PluginResult;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;

public class DIDSessionManagerPlugin extends TrinityPlugin {
    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        try {
            switch (action) {
                case "addIdentityEntry":
                    this.addIdentityEntry(args, callbackContext);
                    break;
                case "deleteIdentityEntry":
                    this.deleteIdentityEntry(args, callbackContext);
                    break;
                case "getIdentityEntries":
                    this.getIdentityEntries(args, callbackContext);
                    break;
                case "getSignedInIdentity":
                    this.getSignedInIdentity(args, callbackContext);
                    break;
                case "signIn":
                    this.signIn(args, callbackContext);
                    break;
                case "signOut":
                    this.signOut(args, callbackContext);
                    break;
                default:
                    return false;
            }
        }
        catch (Exception e) {
            callbackContext.error(e.getLocalizedMessage());
        }
        return true;
    }

    private void sendSuccess(CallbackContext callbackContext, JSONObject jsonObj) {
        callbackContext.sendPluginResult(new PluginResult(PluginResult.Status.OK, jsonObj));
    }

    private void addIdentityEntry(JSONArray args, CallbackContext callbackContext) throws Exception {
        if (args.length() != 1) {
            callbackContext.error("Wrong number of parameters passed");
            return;
        }

        JSONObject identityEntryJson = args.getJSONObject(0);

        DIDSessionManager.IdentityEntry identityEntry = DIDSessionManager.IdentityEntry.fromJsonObject(identityEntryJson);
        DIDSessionManager.getSharedInstance().addIdentityEntry(identityEntry);

        callbackContext.success();
    }

    private void deleteIdentityEntry(JSONArray args, CallbackContext callbackContext) throws Exception {
        if (args.length() != 1) {
            callbackContext.error("Wrong number of parameters passed");
            return;
        }

        String didString = args.getString(0);

        DIDSessionManager.getSharedInstance().deleteIdentityEntry(didString);

        callbackContext.success();
    }

    private void getIdentityEntries(JSONArray args, CallbackContext callbackContext) throws Exception {
        if (args.length() != 0) {
            callbackContext.error("Wrong number of parameters passed");
            return;
        }

        ArrayList<DIDSessionManager.IdentityEntry> entries = DIDSessionManager.getSharedInstance().getIdentityEntries();

        JSONObject jsonObj = new JSONObject();
        JSONArray jsonEntries = new JSONArray();
        for (DIDSessionManager.IdentityEntry entry : entries) {
            jsonEntries.put(entry.asJsonObject());
        }
        jsonObj.put("entries", jsonEntries);

        sendSuccess(callbackContext, jsonObj);
    }

    private void getSignedInIdentity(JSONArray args, CallbackContext callbackContext) throws Exception {
        DIDSessionManager.IdentityEntry signedInIdentity = DIDSessionManager.getSharedInstance().getSignedInIdentity();

        if (signedInIdentity == null)
            callbackContext.success(); // Not signed in, no data to return
        else
            sendSuccess(callbackContext, signedInIdentity.asJsonObject());
    }

    private void signIn(JSONArray args, CallbackContext callbackContext) throws Exception {
        if (args.length() != 1) {
            callbackContext.error("Wrong number of parameters passed");
            return;
        }

        JSONObject identityEntryJson = args.getJSONObject(0);
        DIDSessionManager.IdentityEntry identityToSignIn = DIDSessionManager.IdentityEntry.fromJsonObject(identityEntryJson);
        DIDSessionManager.getSharedInstance().signIn(identityToSignIn);

        callbackContext.success();
    }

    private void signOut(JSONArray args, CallbackContext callbackContext) throws Exception {
        DIDSessionManager.getSharedInstance().signOut();
        callbackContext.success();
    }
}

File: Trinity.Runtime/platform_src/ios/elastOS/Classes/DIDSessionManagerPlugin.swift
/*
 * Copyright (c) 2020 Elastos Foundation
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

public class DIDSessionManagerPlugin : TrinityPlugin {
    func success(_ command: CDVInvokedUrlCommand) {
        let result = CDVPluginResult(status: CDVCommandStatus_OK)
        
        self.commandDelegate.send(result, callbackId: command.callbackId)
    }
    
    func success(_ command: CDVInvokedUrlCommand, _ retAsDict: Dictionary<String, Any>) {
        let result = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: retAsDict)
        self.commandDelegate.send(result, callbackId: command.callbackId)
    }
    
    func error(_ command: CDVInvokedUrlCommand, _ retAsString: String) {
        let result = CDVPluginResult(status: CDVCommandStatus_ERROR,
                                     messageAs: retAsString);
        
        self.commandDelegate.send(result, callbackId: command.callbackId)
    }
    
    func error(_ command: CDVInvokedUrlCommand, _ method: String, _ retAsString: String) {
        let result = CDVPluginResult(status: CDVCommandStatus_ERROR,
                                     messageAs: "\(method): \(retAsString)");
        
        self.commandDelegate.send(result, callbackId: command.callbackId)
    }

    @objc func addIdentityEntry(_ command: CDVInvokedUrlCommand) {
        if command.arguments.count != 1 {
            error(command, "Wrong number of parameters passed")
            return
        }

        let identityEntryJson = command.arguments[0] as! Dictionary<String, Any>

        do {
            if let identityEntry = IdentityEntry.fromJsonObject(identityEntryJson) {
                try DIDSessionManager.getSharedInstance().addIdentityEntry(entry: identityEntry)
                success(command)
            }
            else {
                self.error(command, "addIdentityEntry", "Invalid identity entry format")
            }
        }
        catch let error {
            self.error(command, "addIdentityEntry", error.localizedDescription)
        }
    }

    @objc func deleteIdentityEntry(_ command: CDVInvokedUrlCommand) {
        if command.arguments.count != 1 {
            error(command, "Wrong number of parameters passed")
            return
        }

        do {
            if let didString = command.arguments[0] as? String {
                try DIDSessionManager.getSharedInstance().deleteIdentityEntry(didString: didString)
                success(command)
            }
            else {
                self.error(command, "deleteIdentityEntry", "Invalid DID string")
            }
        }
        catch let error {
            self.error(command, "deleteIdentityEntry", error.localizedDescription)
        }
    }

    @objc func getIdentityEntries(_ command: CDVInvokedUrlCommand) {
        if command.arguments.count != 0 {
            error(command, "Wrong number of parameters passed")
            return
        }

        do {
            let entries = try DIDSessionManager.getSharedInstance().getIdentityEntries()

            var jsonObj = Dictionary<String, Any>()
            var jsonEntries = Array<Dictionary<String, Any>>()
            for entry in entries {
                jsonEntries.append(entry.asJsonObject())
            }
            jsonObj["entries"] = jsonEntries

            success(command, jsonObj)
        }
        catch let error {
            self.error(command, "getIdentityEntries", error.localizedDescription)
        }
    }

    @objc func getSignedInIdentity(_ command: CDVInvokedUrlCommand) {
        do {
            if let signedInIdentity = try DIDSessionManager.getSharedInstance().getSignedInIdentity() {
                success(command, signedInIdentity.asJsonObject())
            }
            else {
                success(command) // Not signed in, no data to return
            }
        }
        catch let error {
            self.error(command, "getSignedInIdentity", error.localizedDescription)
        }
    }

    @objc func signIn(_ command: CDVInvokedUrlCommand) {
        if command.arguments.count != 1 {
            error(command, "signIn", "Wrong number of parameters passed")
            return
        }

        do {
            if let identityEntryJson = command.arguments[0] as? Dictionary<String, Any>,
                let identityToSignIn = IdentityEntry.fromJsonObject(identityEntryJson) {
                
                try DIDSessionManager.getSharedInstance().signIn(identityToSignIn: identityToSignIn)

                success(command)
            }
            else {
                self.error(command, "signIn", "Invalid identity entry format")
            }
        }
        catch let error {
            self.error(command, "signIn", error.localizedDescription)
        }
    }

    @objc func signOut(_ command: CDVInvokedUrlCommand) {
        do {
            try DIDSessionManager.getSharedInstance().signOut()
            success(command)
        }
        catch let error {
            self.error(command, "signOut", error.localizedDescription)
        }
    }
}
